include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp

```diff
@@ -96,89 +105,131 @@ MultiVoxelGridCovariance<PointT> &pclomp::MultiVoxelGridCovariance<PointT>::oper
     *voxel_centroids_ptr_ = *other.voxel_centroids_ptr_;
   }
 
+  removed_count_ = other.removed_count_;
+
   return *this;
 }
 
 template<typename PointT>
 MultiVoxelGridCovariance<PointT> &pclomp::MultiVoxelGridCovariance<PointT>::operator=(MultiVoxelGridCovariance &&other) {
   pcl::VoxelGrid<PointT>::operator=(std::move(other));
-  leaves_ = std::move(other.leaves_);
-  grid_leaves_ = std::move(other.grid_leaves_);
-  leaf_indices_ = std::move(other.leaf_indices_);
-  kdtree_ = std::move(other.kdtree_);
   voxel_centroids_ptr_ = std::move(other.voxel_centroids_ptr_);
+  sid_to_iid_ = std::move(other.sid_to_iid_);
+  grid_list_ = std::move(other.grid_list_);
+  kdtree_ = std::move(other.kdtree_);
+  leaf_indices_ = std::move(other.leaf_indices_);
+
   min_points_per_voxel_ = other.min_points_per_voxel_;
   min_covar_eigvalue_mult_ = other.min_covar_eigvalue_mult_;
 
+  removed_count_ = other.removed_count_;
+
   return *this;
 }
 
 template<typename PointT>
 void MultiVoxelGridCovariance<PointT>::setInputCloudAndFilter(const PointCloudConstPtr &cloud, const std::string &grid_id) {
-  LeafDict leaves;
-  applyFilter(cloud, grid_id, leaves);
+  // Creat a new grid and push it to the back of the vector grid_list_
+  // The vector grid_list_ will be rebuilt after we remove obsolete grids
+  GridNodePtr new_grid(new GridNodeType);
 
-  grid_leaves_[grid_id] = leaves;
+  grid_list_.push_back(new_grid);
+
+  sid_to_iid_[grid_id] = grid_list_.size() - 1;
+
+  applyFilter(cloud, *new_grid);
 }
 
 template<typename PointT>
 void MultiVoxelGridCovariance<PointT>::removeCloud(const std::string &grid_id) {
-  grid_leaves_.erase(grid_id);
+  auto iid = sid_to_iid_.find(grid_id);
+
+  // Set the pointer corresponding to the specified grid to null
+  grid_list_[iid->second].reset();
```

Comment:
I think if a user tries to remove a grid_id that is not kept, it will crash because iid is end() and iid->second is not valid.